### PR TITLE
Go consensus phase 2

### PR DESCRIPTION
**Pull request by Anghul**

### CODE REVIEW

File: be1-go/channel/consensus/mod.go [hunk 1]
```diff
@@ -258,12 +405,321 @@ func (c *Channel) processConsensusElectAccept(data messagedata.ConsensusElectAcc
...
+	if messageState.currentPhase < AcceptPhase {
+		return xerrors.Errorf("consensus corresponding to the message hasn't" +
+			" entered the accept phase")
+	}
+
+	consensusInstance := c.consensusInstances[data.InstanceID]
+	consensusInstance.Lock()
+	defer consensusInstance.Unlock()
+
+	consensusInstance.accepts = append(consensusInstance.accepts, data)
```
Comment (arnauds5) [inline be1-go/channel/consensus/mod.go:685]: Some checks are missing
```
on <ACCEPT, accepted_try, accepted_value> with accepted_try == proposed_try && accepted_value == proposed_value
    Accepts <-  Accepts ∪ {accepted_try, accepted_value} 
```

File: be1-go/channel/consensus/mod.go [hunk 2]
```diff
@@ -258,12 +405,321 @@ func (c *Channel) processConsensusElectAccept(data messagedata.ConsensusElectAcc
...
+
+	if len(consensusInstance.accepts) >= c.hub.GetServerNumber()/2+1 &&
+		messageState.proposer.Equal(c.hub.GetPubKeyOrg()) {
+
+		if !consensusInstance.decided {
+			consensusInstance.decided = true
+			consensusInstance.decision = consensusInstance.proposed_value
```
Comment (arnauds5) [inline be1-go/channel/consensus/mod.go:692]: consensusInstance.proposed_value is never modified and is still false at this point. (Learn always have decision=false here)

**Review by arnauds5 — CHANGES_REQUESTED** [review]

**Review by celinecamacho — COMMENTED** [review]
Looks great. Thank you for all the tests! I left a few comments about the consensus channel.

**Review by rjaccard — APPROVED** [review]
Nice pull request! Very well tested, that's cool :) 
I made a few (not very useful) comments.

File: be1-go/channel/consensus/mod.go [hunk 3]
```diff
@@ -189,84 +239,552 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
...
+
+	return nil
+}
+
+// ProcessConsensusPrepare processes a prepare action.
+func (c *Channel) processConsensusPrepare(msg message.Message) error {
+
+	var data messagedata.ConsensusPrepare
+
+	err := msg.UnmarshalData(&msg)
```
Comment (arnauds5) [inline be1-go/channel/consensus/mod.go:441]: This typo was causing the problem discussed on Slack.
```suggestion
	err := msg.UnmarshalData(&data)
```

**Review by arnauds5 — CHANGES_REQUESTED** [review]

File: be1-go/channel/consensus/mod.go [hunk 4]
```diff
@@ -187,95 +237,637 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
...
+	var data messagedata.ConsensusPrepare
+
+	err := msg.UnmarshalData(&data)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal consensus#prepare: %v", err)
+	}
+
+	if data.Value.ProposedTry == 0 {
+		return xerrors.Errorf("test")
+	}
```
Comment (arnauds5) [inline be1-go/channel/consensus/mod.go:448]: Was this check added only for debuging ? If yes, it can be removed.

**Review by arnauds5 — DISMISSED** [review]

**Review by nkcr — APPROVED** [review]

**Pull request merged**